### ISSUE TITLE
Intermittent "session is already started" error for bluetooth tests

### ISSUE THREAD

**Issue by jdm** [servo/servo#22619]
```
Traceback (most recent call last):
  File "/Users/servo/buildbot/slave/mac-rel-wpt4/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/protocol.py", line 48, in setup
    self.connect()
  File "/Users/servo/buildbot/slave/mac-rel-wpt4/build/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/executors/executorservodriver.py", line 100, in connect
    self.session.start()
  File "/Users/servo/buildbot/slave/mac-rel-wpt4/build/tests/wpt/web-platform-tests/tools/webdriver/webdriver/client.py", line 421, in start
    value = self.send_command("POST", "session", body=body)
  File "/Users/servo/buildbot/slave/mac-rel-wpt4/build/tests/wpt/web-platform-tests/tools/webdriver/webdriver/client.py", line 472, in send_command
    raise err
SessionNotCreatedException: session not created (500): Session is already started


  ▶ ERROR [expected OK] /bluetooth/characteristic/getDescriptors/gen-characteristic-is-removed.https.html
  └   → ''
```

I recall seeing this in #21370 as well.

**Comment by aditj** [comment]
Hey @jdm I was looking at the issue, have you figured out why this is being caused? Why is a [new session](https://github.com/servo/servo/blob/db29cb01b0450c29d0ac45c6de2819a4f87e400a/tests/wpt/web-platform-tests/tools/webdriver/webdriver/client.py#L421) being created when a session already exists? 
Is this because [this check](https://github.com/servo/servo/blob/db29cb01b0450c29d0ac45c6de2819a4f87e400a/tests/wpt/web-platform-tests/tools/webdriver/webdriver/client.py#L413) is not working ? or some other issue perhaps?
 Also is the webdriver ready for multiple sessions?

**Comment by jdm** [comment]
I have not figured out why yet. I believe it should be possible to add support for multiple sessions to the server, though.

**Comment by georgeroman** [comment]
I tried to replicate this error locally but instead of the `SessionNotCreatedException` here is what I've got:
```
[ERROR script::dom::bindings::error] Error at https://web-platform.test:8443/bluetooth/resources/bluetooth-helpers.js:177:10 Illegal constructor.
[ERROR script::dom::bindings::error] Error at https://web-platform.test:8443/bluetooth/characteristic/getDescriptors/gen-characteristic-is-removed.https.html:11:18 Illegal constructor.
```

**Comment by jdm** [comment]
It should be possible to reproduce by running `./mach test-wpt --processes 8 --headless --product=servodriver bluetooth`. We should file a separate issue for the `Illegal constructor` exceptions triggered by https://github.com/web-platform-tests/wpt/blob/e45bda7edddd2cffb4169efe2f61812537b09abd/bluetooth/resources/bluetooth-helpers.js#L177.

**Comment by aditj** [comment]
@jdm , i couldnt reproduce this error all my tests ran as expected , can this error be system specific?

**Comment by georgeroman** [comment]
@aditj Try modifying the test's expectation to `PASS` in https://github.com/servo/servo/blob/master/tests/wpt/metadata/bluetooth/characteristic/getDescriptors/gen-characteristic-is-removed.https.html.ini.

**Comment by aditj** [comment]
> @aditj Try modifying the test's expectation to `PASS` in https://github.com/servo/servo/blob/master/tests/wpt/metadata/bluetooth/characteristic/getDescriptors/gen-characteristic-is-removed.https.html.ini.
Oh !II am now getting the `Illegal constructor` error but not the `SessionNotCreatedException`

**Comment by jgraham** [comment]
> I have not figured out why yet. I believe it should be possible to add support for multiple sessions to the server, though.

Adding multiple sessions to the server probably isn't that easy in the sense that the `webdriver` crate isn't designed to have multiple clients at a time, and ensuring the sessions are isolated on the servo side seems like it would be a lot of work since you need to maintain a map between session id and all the resources associated with that session.

Possibly this just means "shutting down a session and starting a new one without quitting servo", which I think is much more reasonable. However in this case I guess there's some race condition in shutting down the session which allows us to try to start a new one before the previous one is finished.

**Comment by hundredeir** [comment]
As I can see whenever any test function runs it calls a fixture which tries to create a new session and if gets `SessionNotCreated` error then it raises the exception. And between two test functions there isn't any cleanup/teardown happening.
So, when the next function is called the `webdriver` doesn't  even let the message reach servo and itself raises an error.

**Comment by georgeroman** [comment]
I managed to reproduce the error locally and I noticed that two different processes attempted to start WebDriver on the same port:
![servo1](https://user-images.githubusercontent.com/30772943/57395483-e3c53800-71d0-11e9-91c1-a6dc29e97fc5.png)

A few lines down in the log there are two attempts to starting a WebDriver session on the same port:
![servo2](https://user-images.githubusercontent.com/30772943/57395651-5504eb00-71d1-11e9-8485-0318d03fe888.png)

Next, the error was raised. From what I investigated, the ports are chosen [here](https://github.com/servo/servo/blob/master/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/browsers/base.py#L39) and that code is vulnerable to race conditions.

**Comment by jdm** [comment]
That looks extremely suspicious!

**Comment by atouchet** [comment]
Is this issue fixed now?

**Comment by georgeroman** [comment]
The issue is fixed for the bluetooth tests, but it's still there in the new_session wdspec tests.

**Comment by georgeroman** [comment]
So, I further investigated what causes the `Session is already started` error in the `new_session` wdspec tests and I found out that the WebDriver instances that get assigned to each process each have a port of `4444`. Here are some debug prints illustrating this:
![servo](https://user-images.githubusercontent.com/30772943/58353959-a6230900-7e78-11e9-96c6-3e25621573c4.png)

Although the port retrieval in [`_find_next_free_port`](https://github.com/servo/servo/blob/2181872973ce49b50ed2dd077c4fdde6f4c5d019/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L115) is protected with a lock, it must be the case that each process has a different copy of the [`WebDriverServer`](https://github.com/servo/servo/blob/2181872973ce49b50ed2dd077c4fdde6f4c5d019/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L18) class and in this case the lock is irrelevant.

**Comment by georgeroman** [comment]
So, after lots of debugging, I finally managed to understand the real cause of the error in the wdspec `new_session` tests.

When running the wdspec tests in parallel, `wpt` starts new [processes](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py#L498) having as target the [`start_runner`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py#L130) function. Here, the executor is [`ServoWdspecExecutor`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py#L148), which has an associated [`ServoDriverServer`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L204). So, the port initialization occurs for each instance of [`ServoDriverServer`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L204) in a different process. Although [`_used_ports`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L22) is a class variable and is protected by a lock from concurrent access, since we are in different processes, each process will have a different copy of it. In other words, [`_used_ports`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L22) is not shared between the processes (the changes made by one process to it are not visible by the other processes) and the lock that protects it is useless since there is only one process that modifies a particular copy of [`_used_ports`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L22). Because the port initialization is not synchronized we have multiple processes trying to create new sessions on the same `WebDriver` port and so we get those `Session is already started` errors.

Why didn't this also happen in the other occurrences of `get_free_port` ([here for example](https://github.com/georgeroman/wpt/blob/cb9f56a0783391ff991532e930c3bf06eeb972ea/tools/wptrunner/wptrunner/browsers/servodriver.py#L77))? Because there the initialization occurs in [threads](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/testrunner.py#L266), not processes, and each thread shares the same `used_ports` variable. (I've used `multiprocessing.Lock` for protecting `used_ports`, but `threading.Lock` would have also been good enough in this case).

I thought about several possible solutions:
* run the `new_session` tests in a single process
* find a way to share [`_used_ports`](https://github.com/servo/servo/blob/2a61a0dbb02fae6354f76b1abe2fe2ac6c3cb85d/tests/wpt/web-platform-tests/tools/wptrunner/wptrunner/webdriver_server.py#L22) between processes
* modify `wptrunner` to do the port initialization in threads instead of processes
* support multiple sessions in `WebDriver`

**Comment by georgeroman** [comment]
Good news! Once we update `web-platform-tests` to include [this commit](https://github.com/web-platform-tests/wpt/commit/81902b3fc85f89baf1c4473454695680991cf72d), this issue will be fixed!

**Comment by georgeroman** [comment]
This can be closed now!
